### ISSUE TITLE
[REQUEST]: Support C++23 std modules

### ISSUE THREAD

**Issue by ChuanqiXu9** [compiler-explorer/compiler-explorer#5404]
### Is your feature request related to a problem? Please describe

C++23 introduces std modules (std module and std.compat module).  So we can write the following code for C++ in the future:

```
import std;

int main() {
    std::cout << "Hello Modular World!" << std::endl;
}
```

While the progress of modularization is relatively slow, all of the big 3 compilers (gcc, clang and msvc) had offered basic modules support and both libc++ and MSSTL offered experimental std module support now. So technically, we can have a feeling of std module now.

However, different from other C++ features, modules can't be adapted for free for tools like compiler-explorer. Since modules bring new dependent relationship between source files so that the canonical compilation model wouldn't work well for modules. 

### Describe the solution you'd like

As the first step, I think we can try to support the std module for clang and libc++. Given the complexity of modules, we need the involvement of cmake.

The methods to enable std module in libc++ are described in https://libcxx.llvm.org/Modules.html. So basically in CE we can follow the steps:
1. Create a build dir for libc++ under /opt/compiler-explorer. This is required since libc++ devs haven't decide where should be the installation destinations yet.
2. Add `-DLIBCXX_ENABLE_STD_MODULES=ON`  to the cmake argument.
3. When users of compiler-explorer enables cmake, pass `-DLIBCXX_BUILD=<libcxx-build>` to the cmake argument automatically.

Then we are done. I tried the model locally and things look pretty good. The following picture is what I experimented locally.

![image](https://github.com/compiler-explorer/compiler-explorer/assets/68680648/666b471c-d11b-49d5-9bd4-d269879d2a20)

### Describe alternatives you've considered

@Quuxplusone said he doesn't like the idea since he doesn't want the users to learn the edge-cutting cmake feature. Then he suggested to find a method to generate the cmake scripts automatically.

@mathstuf suggested to write a tool to generate and pass all the needed module files automatically.

While I think both the suggestion may be implementable due to the scale of the specific issue, I think the don't necessarily need to be the first step. I feel simpler the method is, the better it is.

**Comment by Quuxplusone** [comment]
> @Quuxplusone said he doesn't like the idea since he doesn't want the users to learn the [cutting-edge] cmake feature.

Right. I'm willing to offer up the "Clang experimental P1144" branch as a guinea pig for Modules-related interface ideas, (because/but) what I'm really interested in is an outcome where newbies can go to https://godbolt.org/z/584deWjhG , type

    import std;
    int main() {
        std::print("Hello world\n");
    }

and have it Just Work. I'm _not_ motivated by outcomes where the user has to open multiple panes, some of which aren't even C++, and memorize things to type into them. That CMake stuff can still _happen_, but it needs to be "behind the scenes." (That is: "He suggested to find a method to generate the cmake scripts automatically.")  Alternatively, all that wiring can be exposed for the benefit of bleeding-edge users, but then _I'm_ not motivated to explore it, because exposed wiring won't get me where _I'm_ interested in going.

@JohelEGP points out that this CMake-based solution requires recompiling the entire Standard Library for every "hello world," because a Modules .pcm file (unlike a library .o/.a file) is sensitive to _all_ compiler options. IIUC you can't just build the library once and then "link against the .pcm file"; you need to rebuild the library's .pcm file every time the command-line options change. I suggested maybe we could precompile some common sets of arguments (`-O0`, `-O2`, `-O3`) in the image to avoid this cost... _sometimes_. If "sometimes recompile the entire library" is an _unavoidable_ cost, then Modules hello-worlds will be even slower than Ranges hello-worlds, and then it won't seem reasonable to pursue Modules _at all_ IMO.

**Comment by JohelEGP** [comment]
> @JohelEGP points out that this CMake-based solution requires recompiling the entire Standard Library for every "hello world," because a Modules .pcm file (unlike a library .o/.a file) is sensitive to _all_ compiler options.

By the way, compiling `std.cppm` now takes 5 s on my system, and occupies 30 MB.

Based on my experience comparing CE build times to local build times,
I'd say that could sometimes takes 10 s to rebuild in CE.
That takes a lot from the 20 s of build time you get until the timeout.

That said, I know there are efforts in the space that will further improve this, including:
- Even lighter-weight module builds: <https://discourse.llvm.org/t/rfc-c-20-modules-introduce-thin-bmi-and-decls-hash/74755>.
- Reducing the need for rebuild by recognizing many flags that won't actually affect the build output.

**Comment by mathstuf** [comment]
> Reducing the need for rebuild by recognizing many flags that won't actually affect the build output.

Daniel Ruoso published a WG21 paper about getting a "compatibility hash" for a given command line.

**Comment by ChuanqiXu9** [comment]
> Even lighter-weight module builds: https://discourse.llvm.org/t/rfc-c-20-modules-introduce-thin-bmi-and-decls-hash/74755.

BTW, the so called thin-bmi may not reduce the size of std modules significantly. Since almost all the entities in std.cppm are templates or inline.

**Comment by ruoso** [comment]
> Daniel Ruoso published a WG21 paper about getting a "compatibility hash" for a given command line

Yes, that's [P2581R2](https://wg21.link/P2581R2). The work-around until compilers support that is to just checksum the "baseline compile command" itself, which is not going to be as useful as if the compiler did it, but it's still going to have some amount of reuse. 

See also [P2962R0](https://wg21.link/P2962R0) for more discussions on "baseline compile command".

**Comment by poyaoc97** [comment]
Now that std and std.compat modules from libc++ can be installed (https://github.com/llvm/llvm-project/commit/8b47bb657b5905d954b9041415020358802407d5, made it to the LLVM 18 train), can we enable `LIBCXX_INSTALL_MODULES_DIR` so that we can access the module at `/opt/compiler-explorer/clang-trunk/share/libc++/v1/std.cppm` before build systems are ready?

**Comment by partouf** [comment]
> Now that std and std.compat modules from libc++ can be installed ([llvm/llvm-project@8b47bb6](https://github.com/llvm/llvm-project/commit/8b47bb657b5905d954b9041415020358802407d5), made it to the LLVM 18 train), can we enable `LIBCXX_INSTALL_MODULES_DIR` so that we can access the module at `/opt/compiler-explorer/clang-trunk/share/libc++/v1/std.cppm` before build systems are ready?

that's not a particular logical path, unless the cppm is somehow architecture insensitive?

**Comment by ruoso** [comment]
> that's not a particular logical path, unless the cppm is somehow architecture insensitive?

Those are source files, not binaries. The build system will need to produce the BMI for those modules.

**Comment by poyaoc97** [comment]
Looks like
```diff
+         CMAKE_EXTRA_ARGS+=("-DCLANG_ENABLE_HLSL=On")
-         CMAKE_EXTRA_ARGS+=("-DCLANG_ENABLE_HLSL=On" "-DLIBCXX_INSTALL_MODULES=ON")
```
at https://github.com/compiler-explorer/clang-builder/blob/8b4f92d7cbad398b18a6f7d74b59aa2b5eff68b9/build/build.sh#L190 would do.

**Comment by bretbrownjr** [comment]
I think that's the first step.

In addition to that, the build system, Compiler Explorer, and/or the compiler will need to guess how to parse the std and std.compat modules in a way that is compatible enough with the source files importing the standard modules. Especially flags like the C++ standard version need to match.

Probably the compiler can figure out how to juggle BMIs (Built Module Interface files), but everything else will need to manage those intermediate files as well.

BMI management for the std module isn't implemented for CMake yet.

**Comment by ChuanqiXu9** [comment]
It would be pretty good if we can use `import std;` in compiler-explorer. In my experience, I found many people don't know libc++ already has experimental std modules support. It can be greatly improved if compiler-explorer can support that.

But I also understand that this is not super trivial for compiler-explorer. In the most use cases, the compiler-explorer don't invoke build systems. There is only a single file mode. I am wondering if it is too heavy to use CMake for std module and std.compat module only for the default single file mode. 

My idea for compiler-explorer is to read https://clang.llvm.org/docs/StandardCPlusPlusModules.html and write a script itself to manage things.

For example, if C++20 is enabled (std module is backported in C++20 in implementations) and no build system is involved (in the single file mode, the default mode), the compiler-explorer can call `clang-scan-deps` to scan the sources, and if there is only std module or std.compat module, the compiler-explorer can build std.pcm and std.compat.pcm according to the flags for the single file and pass that to the compilation job of the single file. (If there are more modules, the compiler-explorer can emit an error directly.)

If the users want to import other modules, they have to invoke build systems. In such cases, the compiler-explorer don't need to do a lot of works except invoking the new build systems.

There are other question about how to manage the cache since the size of BMI is larger than sources. But we don't need to talk about this in the early stage.

**Comment by partouf** [comment]
Ok, so with the PR merged and built we now have this:

```
$ ls -l /opt/compiler-explorer/clang-trunk/share/libc++/v1
total 28
drwxr-xr-x 2 root root 6144 Feb 17 00:34 std
drwxr-xr-x 2 root root 6144 Feb 17 00:34 std.compat
-rw-r--r-- 1 root root 4639 Feb 17 00:33 std.compat.cppm
-rw-r--r-- 1 root root 9552 Feb 17 00:33 std.cppm
```

Where to go from here...

I gave the example CMake a try, but am getting an error: https://compiler-explorer.com/z/Y68hrPh9T

**Comment by poyaoc97** [comment]
Looks like you need to specify the standard in CMake file explicitly instead of just on the compiler pane.
Like so: https://godbolt.org/z/f9eKWEn84

**Comment by partouf** [comment]
I guess that makes sense.
Nice! Thanks!

**Comment by TheRustifyer** [comment]
> Looks like you need to specify the standard in CMake file explicitly instead of just on the compiler pane. Like so: https://godbolt.org/z/f9eKWEn84

So finally, is needed to use a CMake script to have `std` or `std.compat`?
Are still pllaned tomaking it usable directly on single file (enable standard modules) from the compiler explorer panel?

Thanks

**Comment by partouf** [comment]
> So finally, is needed to use a CMake script to have `std` or `std.compat`? Are still pllaned tomaking it usable directly on single file (enable standard modules) from the compiler explorer panel?
> 
> Thanks

There's not much chance of that happening this year. Maybe once both GCC and Clang have a solid solution and stable API + ABI for it and everything's non-experimental, but even then it might turn out to be too complicated.

**Comment by bretbrownjr** [comment]
Toolchain support for `import std;` is still maturing. It is wise to engage directly with your favorite toolchain(s) to make sure there's a well understood way to get the std modules to work. It's more complicated than you might expect, especially given compilers (like clang) can support several versions of different standard libraries, so even hardcoding assumptions can be more involved than you might expect.

There's some chance that the best initial use of `import std;` in Compiler Explorer is to share toolchain bugs in getting it to work, especially when standard library includes are mixed in.

There's also some chance that toolchain support for `import std;` will have some pretty significant limitations. I wouldn't be surprised if we all end up using CMake boilerplate more often in our Compiler Explorer examples at some point.

**Comment by Quuxplusone** [comment]
@bretbrownjr: I can't speak for @TheRustifyer, but FWIW, _my_ interest in Modules-support-on-CE remains entirely the super newbie friendly version. I don't want to mess with CMake or multiple files or anything. (I very much understand that that's useful especially for sharing bugs; I'm not saying that I want the CMake stuff to be impossible! I just want this different thing _too_, somehow. Probably as a completely different "compiler" in the dropdown.) I want to be able to share a link to a godbolt containing
```
import std;
int main() {
  std::println("hello world");
}
```
and show that it compiles cleanly and runs and prints. I know CE would have to be doing something really complicated and bleeding-edge and not-at-all-"well understood" in the background... but that's the simple _foreground_ experience I'd want to be able to show people with Modules.

**Comment by bretbrownjr** [comment]
There's a good chance that Compiler Explorer would end up creating (and hiding?) a build system configuration. Modules require extra compiler interaction and therefore more compiler commands.

It's not impossible that clang, MSVC, EDG, and GCC each support online commands that support the std module. It could even be likely on a long enough time horizon.

But there's a good chance that the simplest available technology is some CMake boilerplate somewhere, especially if you want more than one toolchain building a given example.

Possibly someone volunteers to implement and/or what you want in all the compilers? It hasn't exactly happened yet, but it could happen. It might actually be simpler to dramatically cut down on the amount of needed CMake boilerplate via a oneliner module that understands pitchfork project structure or something.

**Comment by ChuanqiXu9** [comment]
It is understandable that we can't proceed quickly due to complexity. But probably we **have to** invent something within compiler-explorer for std modules.

For example, a hello world example with std modules:

```
import std;
int main() {
    std::cout << "Hello World.\n";
}
```

takes `4.1s` to compile in my local machine, where the std module itself needs `3.9s` to compile and the code it self only takes `0.091s` to compile. This may be fine for a lot of projects. Since the std module is rarely change for most people. So we can avoid the recompilation of std modules as long as we don't run `ninja clean` in the projects.

However, it is a different story for the online services like compiler-explorer. The `4s` can heavily repeat a lot of times with the invocations of compiler-explorer all over the world. This may be some terrible user experience and not simple to solve..
Cache management is always not easy..

(Maybe we can use ccache to build std modules? But note that the ccache don't support modules yet...)

**Comment by mathstuf** [comment]
Once CMake has a usage hash, there may be a way to use the hash CMake uses for the `.bmi` names of imported modules with some cache.

**Comment by ChuanqiXu9** [comment]
> Once CMake has a usage hash, there may be a way to use the hash CMake uses for the `.bmi` names of imported modules with some cache.

Out of curiousity, is there any link to it? I am a little bit surprised since I didn't expect the cmake to reuse compilation product across projects (except the installed things on the system, but we've decided not to install the BMI..)

**Comment by rdong8** [comment]
CMakeLists.txt:

```cmake
cmake_minimum_required(VERSION 3.30.0 FATAL_ERROR)

set(CMAKE_EXPERIMENTAL_CXX_IMPORT_STD "0e5b6991-d74f-4b3d-a41c-cf096e0b2508")
set(CMAKE_CXX_MODULE_STD ON)
set(CMAKE_CXX_SCAN_FOR_MODULES ON)

project(example LANGUAGES CXX)

set(CMAKE_CXX_STANDARD "${CMAKE_CXX_STANDARD_LATEST}")
set(CMAKE_CXX_STANDARD_REQUIRED True)

add_executable(example)
target_sources(example
  PRIVATE FILE_SET CXX_MODULES FILES
  main.ccm
)
```

example.ccm:

```cpp
export module example;

import std;

auto main() -> int
{
        std::println("Hello World");
}
```

Commands:

```bash
# Configure
cmake \
        -S . \
        -B build \
        -G "Ninja Multi-Config" \
        -DCMAKE_CXX_COMPILER=clang++ \
        -DCMAKE_CXX_FLAGS=-stdlib=libc++

# Build
cmake \
        --build build \
        --config Debug \
        -t all \
        -j

# Run
build/Debug/example
# Output: Hello World
```

[example.zip](https://github.com/user-attachments/files/17788808/example.zip)

**Comment by liviurazorea** [comment]
`cmake_minimum_required(VERSION 3.30.0 FATAL_ERROR)` fail for me at [link](https://godbolt.org/z/r17s9h7zG). I get 

```
Step cmake returned: 1
-- Configuring incomplete, errors occurred!
CMake Error at CMakeLists.txt:1 (cmake_minimum_required):
  CMake 3.30.0 or higher is required.  You are running version 3.28.0
```

**Comment by rdong8** [comment]
Yeah unfortunately seems like the godbolt CMake version is too low. If you run locally it should work with the [Snap CMake](https://snapcraft.io/cmake).

**Comment by mattgodbolt** [comment]
See also #1892 ; though that's module support more generally.

**Comment by mattgodbolt** [comment]
The cmake on the site is now 3.31.5 - can you re-check and see whether this helps?

**Comment by narpfel** [comment]
Seems to work: https://godbolt.org/z/zWTW8n8fb

... but it doesn’t really work well: CMake has to recompile the standard library module on each recompile. But that’s not really fixable without sharing the built module between compilations (and checking whether the compiler flags are compatible/the same).

**Comment by pavlix** [comment]
I wonder why everybody's talking about cmake here. So far I got modules working with *gcc* and *clang* with just two lines of code. I understand that *cmake* is popular and useful. However, this discussion makes an impression that it's somehow required to make things work. I admit that I found it much easier to experiment  make it work without it.

One can experiment with these comands just in *shell* or can create a fairly trivial *Makefile* to make sure to rebuild the standard library module as needed. Or of course this could be done with any software pipeline that knows when it needs to be rebuilt based on the command-line options as was mentioned already.

GCC:

```
time g++ -std=c++26 -fmodules -fmodule-only -fsearch-include-path bits/std.cc -c
time g++ -std=c++26 -fmodules example.cpp -o example
```

CLang:

```
time time clang++ -std=c++26 -stdlib=libc++ -fprebuilt-module-path=. -Wno-reserved-module-identifier --precompile /usr/share/libc++/v1/std.cppm -o std.pcm
time clang++ -std=c++26 -stdlib=libc++ -fprebuilt-module-path=. example.cpp -o example
```

Note: Of course you can use `-std=c++23` and not all of the command line options are required. This is basicaly taken from my experimental make files for working with modules. I can include shortened versions these as well. It works on *Fedora 43* with downgraded versions of the compiler and standard library due to a bug that might well be already resolved at the time you read this comment.

GCC:

```
CXX := g++
CXXFLAGS := -std=c++26 -fmodules
MY_TARGETS := importstd

all: $(MY_TARGETS)
clean:
	rm -f $(MY_TARGETS)
	rm -f *.o
	rm -f .module.*
	rm -rf gcm.cache
importstd: ../importstd.cpp .module.std
	time $(CXX) $(CXXFLAGS) $< -o $@
.module.std:
	time $(CXX) $(CXXFLAGS) -fmodule-only -fsearch-include-path bits/std.cc -c
	touch $@
```

CLang:

```
CXX := clang++
CXXFLAGS := -std=c++26 -stdlib=libc++ -fprebuilt-module-path=.
MY_TARGETS := importstd importstdheader importmodule importheaderlib

all: $(MY_TARGETS)
clean:
	rm -f $(MY_TARGETS)
	rm -f *.o
	rm -f *.pcm
importstd: ../importstd.cpp std.pcm
	time $(CXX) $(CXXFLAGS) $< -o $@
std.pcm: /usr/share/libc++/v1/std.cppm
	time $(CXX) $(CXXFLAGS) -Wno-reserved-module-identifier --precompile $< -o $@
```

I might have missed something but I don't really see the problem here. It's pretty quick to build and thus I believe it isn't all that bad to have it rebuilt upon each change of compiler & compiler flags + ideally do some caching to save even more compilation time.

**Comment by ChuanqiXu9** [comment]
> I wonder why everybody's talking about cmake here. So far I got modules working with _gcc_ and _clang_ with just two lines of code. I understand that _cmake_ is popular and useful. However, this discussion makes an impression that it's somehow required to make things work. I admit that I found it much easier to experiment make it work without it.
> 
> One can experiment with these comands just in _shell_ or can create a fairly trivial _Makefile_ to make sure to rebuild the standard library module as needed. Or of course this could be done with any software pipeline that knows when it needs to be rebuilt based on the command-line options as was mentioned already.
> 
> GCC:
> 
> ```
> time g++ -std=c++26 -fmodules -fmodule-only -fsearch-include-path bits/std.cc -c
> time g++ -std=c++26 -fmodules example.cpp -o example
> ```
> 
> CLang:
> 
> ```
> time time clang++ -std=c++26 -stdlib=libc++ -fprebuilt-module-path=. -Wno-reserved-module-identifier --precompile /usr/share/libc++/v1/std.cppm -o std.pcm
> time clang++ -std=c++26 -stdlib=libc++ -fprebuilt-module-path=. example.cpp -o example
> ```
> 
> Note: Of course you can use `-std=c++23` and not all of the command line options are required. This is basicaly taken from my experimental make files for working with modules. I can include shortened versions these as well. It works on _Fedora 43_ with downgraded versions of the compiler and standard library due to a bug that might well be already resolved at the time you read this comment.
> 
> GCC:
> 
> ```
> CXX := g++
> CXXFLAGS := -std=c++26 -fmodules
> MY_TARGETS := importstd
> 
> all: $(MY_TARGETS)
> clean:
> 	rm -f $(MY_TARGETS)
> 	rm -f *.o
> 	rm -f .module.*
> 	rm -rf gcm.cache
> importstd: ../importstd.cpp .module.std
> 	time $(CXX) $(CXXFLAGS) $< -o $@
> .module.std:
> 	time $(CXX) $(CXXFLAGS) -fmodule-only -fsearch-include-path bits/std.cc -c
> 	touch $@
> ```
> 
> CLang:
> 
> ```
> CXX := clang++
> CXXFLAGS := -std=c++26 -stdlib=libc++ -fprebuilt-module-path=.
> MY_TARGETS := importstd importstdheader importmodule importheaderlib
> 
> all: $(MY_TARGETS)
> clean:
> 	rm -f $(MY_TARGETS)
> 	rm -f *.o
> 	rm -f *.pcm
> importstd: ../importstd.cpp std.pcm
> 	time $(CXX) $(CXXFLAGS) $< -o $@
> std.pcm: /usr/share/libc++/v1/std.cppm
> 	time $(CXX) $(CXXFLAGS) -Wno-reserved-module-identifier --precompile $< -o $@
> ```
> 
> I might have missed something but I don't really see the problem here. It's pretty quick to build and thus I believe it isn't all that bad to have it rebuilt upon each change of compiler & compiler flags + ideally do some caching to save even more compilation time.

I think the problem is, we can't assume all the TU will import std. They may not import std and they may import std.compat. So there need to be more conditions. But I agree this may not be super hard.

**Comment by ajasmin** [comment]
> I think the problem is, we can't assume all the TU will import std. They may not import std and they may import std.compat. So there need to be more conditions. But I agree this may not be super hard.

Would there be some interest in a small clang-ast based executable that checks if `import std` is used? Though, it might be easier to just grep for the string. Or, for that matter, just let the compiler fail...

The other difficulty, as already mentioned, is caching.

**Comment by ChuanqiXu9** [comment]
> > I think the problem is, we can't assume all the TU will import std. They may not import std and they may import std.compat. So there need to be more conditions. But I agree this may not be super hard.
> 
> Would there be some interest in a small clang-ast based executable that checks if `import std` is used? Though, it might be easier to just grep for the string. Or, for that matter, just let the compiler fail...
> 
> The other difficulty, as already mentioned, is caching.

We have clang-scan-deps for it.

**Comment by mathstuf** [comment]
> I think the problem is, we can't assume all the TU will import std. They may not import std and they may import std.compat. So there need to be more conditions. But I agree this may not be super hard.

You also cannot assume that all TUs will have BMI-compatible flags. If one TU uses `-std=gnu++23`, it needs a separate BMI for `import std;` than one using `-std=c++23`. There are a *lot* of flags for which this is true.

**Comment by Alunderin** [comment]
For anyone who may stumble onto this thread, here's an updated example of what @narpfel wrote with support for both Clang and GCC: <https://godbolt.org/z/9zchhs9xY>

It still recompiles the standard library for every recompilation, which isn't ideal to say the least.

**Comment by rdong8** [comment]
With GCC 16 there's also an even easier way: https://godbolt.org/z/aE5fGPz8T

**Comment by narpfel** [comment]
`clang-trunk` also supports building the standard library module via `-fmodules-driver` (https://github.com/llvm/llvm-project/pull/194475): https://godbolt.org/z/xPezrjvrh

However, this only works for execution and “Link to binary”, otherwise clang complains about CE’s implicit `-o` argument because it wants to generate `std.o` in addition to the expected `example.o` object file.

**Comment by naveen-seth** [comment]
> It still recompiles the standard library for every recompilation, which isn't ideal to say the least.

Hi, I'm currently working on fixing this. It will soon only compile the `.pcm` and use the system provided libary.

I was trying to create a `-fmodules-driver` example with multiple module inputs (and also Clang modules) on the command-line.
Is there some way to disable the `example.cpp` from implicitly being passed?
I feel like it can be confusing if all inputs have to be specified on the command line except one.

https://godbolt.org/z/W7G1v5bdn

**Comment by partouf** [comment]
> > It still recompiles the standard library for every recompilation, which isn't ideal to say the least.
> 
> Hi, I'm currently working on fixing this. It will soon only compile the `.pcm` and use the system provided libary.
> 
> I was trying to create a `-fmodules-driver` example with multiple module inputs (and also Clang modules) on the command-line. Is there some way to disable the `example.cpp` from implicitly being passed? I feel like it can be confusing if all inputs have to be specified on the command line except one.
> 
> https://godbolt.org/z/W7G1v5bdn

the recommended way to use modules is to use cmake, your example doesn't use cmake

**Comment by naveen-seth** [comment]
One of the major motivators for adding `-fmodules-driver` to Clang was to allow for such simple module examples without the use of an external build system.

The issue I have with `example.cpp` being passed implicitly isn’t modules- or `-fmodules-driver`-related. If `example.cpp` isn’t listed in the tree-view “included files” section and isn't specified on the command-line either, the compiler shouldn’t complain that it’s missing imo.
(I think this applies to all command-lines with multiple file inputs).

> However, this only works for execution and “Link to binary”, otherwise clang complains about CE’s implicit -o argument because it wants to generate std.o in addition to the expected example.o object file.

I will also try to fix this soon too (and do quite some work on other bug fixes and the documentation!), and then `-fmodules-driver` might also be a good option for simple modules-use on compiler-explorer.

**Comment by ChuanqiXu9** [comment]
> One of the major motivators for adding `-fmodules-driver` to Clang was to allow for such simple module examples without the use of an external build system.
> 
> The issue I have with `example.cpp` being passed implicitly isn’t modules- or `-fmodules-driver`-related. If `example.cpp` isn’t listed in the tree-view “included files” section and isn't specified on the command-line either, the compiler shouldn’t complain that it’s missing imo. (I think this applies to all command-lines with multiple file inputs).
> 
> > However, this only works for execution and “Link to binary”, otherwise clang complains about CE’s implicit -o argument because it wants to generate std.o in addition to the expected example.o object file.
> 
> I will also try to fix this soon too (and do quite some work on other bug fixes and the documentation!), and then `-fmodules-driver` might also be a good option for simple modules-use on compiler-explorer.

It will be great and useful in practice if https://godbolt.org/z/TTThGqvhz can work.
